components/scream/src/share/grid/remap/coarsening_remapper.cpp

```diff
@@ -0,0 +1,951 @@
+            v(lid,idim) += buf (offset + lidpos*ndims + idim);
+          });
+        });
+      } break;
+      case LayoutType::Scalar3D:
+      {
+        auto v = f.get_view<Real**>();
+        const int nlevs = fl.dims().back();
+        auto policy = ESU::get_default_team_policy(m_total_num_recv_gids,nlevs);
+        Kokkos::parallel_for(policy,
+                             KOKKOS_LAMBDA(const MemberType& team){
+          const int i = team.league_rank();
+          const int lid = lids_pids(i,0);
+          const int pid = lids_pids(i,1);
+          const int lidpos = i - pid_lid_start(pid);
+          const int offset = f_pid_offsets(pid);
+
+          Kokkos::parallel_for(Kokkos::TeamThreadRange(team,nlevs),
+                               [&](const int ilev) {
+            v(lid,ilev) += buf (offset + lidpos*nlevs + ilev);
+          });
+        });
+      } break;
+      case LayoutType::Vector3D:
+      {
+        auto v = f.get_view<Real***>();
+        const int ndims = fl.dim(1);
+        const int nlevs = fl.dims().back();
+        auto policy = ESU::get_default_team_policy(m_total_num_recv_gids,nlevs*ndims);
+        Kokkos::parallel_for(policy,
+                             KOKKOS_LAMBDA(const MemberType& team){
+          const int i = team.league_rank();
+          const int lid = lids_pids(i,0);
+          const int pid = lids_pids(i,1);
+          const int lidpos = i - pid_lid_start(pid);
+          const int offset = f_pid_offsets(pid);
+
+          Kokkos::parallel_for(Kokkos::TeamThreadRange(team,nlevs*ndims),
+                               [&](const int idx) {
+            const int idim = idx / nlevs;
+            const int ilev = idx % nlevs;
+            v(lid,idim,ilev) += buf (offset + lidpos*ndims*nlevs + idim*nlevs + ilev);
+          });
+        });
+      } break;
+
+      default:
+        EKAT_ERROR_MSG ("Unexpected field rank in CoarseningRemapper::pack.\n"
+            "  - MPI rank  : " + std::to_string(m_comm.rank()) + "\n"
+            "  - field rank: " + std::to_string(fl.rank()) + "\n");
+    }
+  }
+}
+
+
+auto CoarseningRemapper::
+get_my_triplets_gids (const std::string& map_file) const
+  -> view_1d<gid_t>::HostMirror
+{
+  using namespace ShortFieldTagsNames;
+
+  scorpio::register_file(map_file,scorpio::FileMode::Read);
+  // Create a "fake" grid, with as many dofs as the number of triplets in the src file
+  const int ngweights = scorpio::get_dimlen_c2f(map_file.c_str(),"n_s");
+  const auto io_grid_linear = create_point_grid ("helper",ngweights,1,m_comm);
+  const int nlweights = io_grid_linear->get_num_local_dofs();
+  const auto& raw_comm = m_comm.mpi_comm();
+
+  gid_t offset = nlweights;
+  m_comm.scan(&offset,1,MPI_SUM);
+  offset -= nlweights; // scan is inclusive, but we need exclusive
+
+  // Read chunk of triplets col indices
+  std::vector<gid_t> cols(nlweights);
+  scorpio::get_variable(map_file, "col", "col", {"n_s"}, "int", "int-nnz");
```

Comment:
So that we can reuse PIO decomps, we should make tag names unique to a decomp or else explicitly deallocate them. Here, "int-nnz" is associated with a different decomp than "int-nnz" later in this routine.

Reply:
Yes, good point. I was indeed getting an issue with decomps getting reused, which I fixed by closing the file every time. But in view of decomps no longer being freed at file closure, I need to make them unique.